### PR TITLE
[RFC] osd: EC Partial Stripe Reads (Retry of #23138)

### PR DESCRIPTION
**Pull request by markhpc**
This is a re-implementation of PR #23138 rebased on main with a couple of nitpicky changes to make the code a little more clear (to me at least).  Credit goes to Xiaofei Cui [cuixiaofei@sangfor.com.cn](mailto:cuixiaofei@sangfor.com.cn) for the original implementation.

Looking at the original PR's review, it does not appear that we can use the same technique as in https://github.com/ceph/ceph/commit/468ad4b41010488c8d48ef65ccbebfdb4270690f.  We don't have the ReadOp yet.  I'm not sure if @gregsforytwo's idea to query the plugin works, but it's clear we are not doing the efficient thing from the get-go here.

The performance and efficiency benefits for small random reads appears to be quite substantial, especially for large stripe widths.

FIO 4KB Random Read IOPS (16 Client)
FIO 4KB Random Read Cycles_OP (16 Client)

Edit: _There was previously a bug in the cycles/op calculation due to a change in how the parsing code works (we previous didn't collect perf data on all osds, now we do).  The scale is exactly the same, but the cycles/op numbers were originally over-inflated by a static factor of 6.  The new numbers have been verified to be within about 10% of the expected cycles/op numbers when calculated using aggregate average CPU consumption and IOPS instead of aggregate cycles and ops._

## Contribution Guidelines
- To sign and title your commits, please refer to [Submitting Patches to Ceph](https://github.com/ceph/ceph/blob/main/SubmittingPatches.rst).

- If you are submitting a fix for a stable branch (e.g. "pacific"), please refer to [Submitting Patches to Ceph - Backports](https://github.com/ceph/ceph/blob/master/SubmittingPatches-backports.rst) for the proper workflow.

## Checklist
- Tracker (select at least one)
  - [ ] References tracker ticket
  - [ ] Very recent bug; references commit where it was introduced
  - [x] New feature (ticket optional)
  - [ ] Doc update (no ticket needed)
  - [x] Code cleanup (no ticket needed)
- Component impact
  - [ ] Affects [Dashboard](https://tracker.ceph.com/projects/dashboard/issues/new), opened tracker ticket
  - [ ] Affects [Orchestrator](https://tracker.ceph.com/projects/orchestrator/issues/new), opened tracker ticket
  - [x] No impact that needs to be tracked
- Documentation (select at least one)
  - [ ] Updates relevant documentation
  - [x] No doc update is appropriate
- Tests (select at least one)
  - [ ] Includes [unit test(s)](https://docs.ceph.com/en/latest/dev/developer_guide/tests-unit-tests/)
  - [ ] Includes [integration test(s)](https://docs.ceph.com/en/latest/dev/developer_guide/testing_integration_tests/)
  - [ ] Includes bug reproducer
  - [x] No tests

<details>
<summary>Show available Jenkins commands</summary>

- `jenkins retest this please`
- `jenkins test classic perf`
- `jenkins test crimson perf`
- `jenkins test signed`
- `jenkins test make check`
- `jenkins test make check arm64`
- `jenkins test submodules`
- `jenkins test dashboard`
- `jenkins test dashboard cephadm`
- `jenkins test api`
- `jenkins test docs`
- `jenkins render docs`
- `jenkins test ceph-volume all`
- `jenkins test ceph-volume tox`
- `jenkins test windows`
</details>

### CODE REVIEW

**Comment by markhpc** [comment]
retest this please

File: src/erasure-code/ErasureCode.cc [hunk 1]
```diff
@@ -332,17 +332,40 @@ int ErasureCode::decode_concat(const map<int, bufferlist> &chunks,
...
+      if (want_to_read.find(i->first) == want_to_read.end()) {
+        need_decode = true;
+        break;
+      }
+      decode_chunks.insert(i->first);
+    }
+    if (!need_decode) {
+      want_to_read.swap(decode_chunks);
+    }
+  }
```
Comment (ronen-fr) [inline src/erasure-code/ErasureCode.cc:355]: Suggested:

  set<int> want_to_read;
  for (unsigned int i = 0; i < get_data_chunk_count(); i++) {
    want_to_read.insert(chunk_index(i));
  }

  bool need_decode = false;
  set<int> decode_chunks;
  if (chunks.size() < get_data_chunk_count()) {

    // for partial_read
    for (const auto& [thing_key, thing_bl] : chunks) {
      // note that we do use c++20
      if (want_to_read.contains(thing_key])) {
        decode_chunks.insert(i->first);
      } else {
        need_decode = true;
        break;
      }
    }

    if (!need_decode) {
      want_to_read.swap(decode_chunks);
    }

File: src/erasure-code/ErasureCode.cc [hunk 2]
```diff
@@ -332,17 +332,40 @@ int ErasureCode::decode_concat(const map<int, bufferlist> &chunks,
...
   map<int, bufferlist> decoded_map;
   int r = _decode(want_to_read, chunks, &decoded_map);
   if (r == 0) {
     for (unsigned int i = 0; i < get_data_chunk_count(); i++) {
-      decoded->claim_append(decoded_map[chunk_index(i)]);
+      if (decoded_map.find(chunk_index(i)) != decoded_map.end()) {
```
Comment (ronen-fr) [inline src/erasure-code/ErasureCode.cc:360]: contains()

File: src/erasure-code/ErasureCode.cc [hunk 3]
```diff
@@ -332,17 +332,40 @@ int ErasureCode::decode_concat(const map<int, bufferlist> &chunks,
...
+        decoded->claim_append(decoded_map[chunk_index(i)]);
+      }
     }
   }
   return r;
 }
+
+bool ErasureCode::is_systematic() {
```
Comment (ronen-fr) [inline src/erasure-code/ErasureCode.cc:368]: should probably be a cost function

File: src/osd/ECBackend.cc [hunk 4]
```diff
@@ -1686,6 +1686,29 @@ int ECBackend::get_min_avail_to_read_shards(
...
+
+  int total_chunks = (chunk_size - 1 + len) / chunk_size;
+  int first_chunk = (off / chunk_size) % data_chunk_count;
+
+  if (total_chunks > data_chunk_count) {
+    total_chunks = data_chunk_count;
+  }
+
+  for(int i = 0; i < total_chunks; i++) {
+    int j = (first_chunk + i) % data_chunk_count;
```
Comment (ronen-fr) [inline src/osd/ECBackend.cc:1706]: please add a comment to explain the logic here (and maybe rename 'j'?)

File: src/osd/ECBackend.cc [hunk 5]
```diff
@@ -2485,14 +2579,30 @@ int ECBackend::send_all_remaining_reads(
     dout(10) << __func__ << " want attrs again" << dendl;
   }
 
+  bool partial_read = rop.to_read.find(hoid)->second.partial_read;
+  // realign the offset and len to make partial reads normal reads.
+  if (partial_read) {
+    list<ec_align_t> new_to_read;
+    for(auto read : to_read) {
```
Comment (ronen-fr) [inline src/osd/ECBackend.cc:2586]: wouldn't 'const auto&' work here?

File: src/erasure-code/ErasureCode.cc (hunk 1 above)
Comment (markhpc) [reply src/erasure-code/ErasureCode.cc:354]: Nice cleanup!

File: src/osd/ECBackend.cc (hunk 4 above)
Comment (markhpc) [reply src/osd/ECBackend.cc:1706]: left the "j" for now. We have a similar loop in get_want_to_read_shards here:
https://github.com/ceph/ceph/blob/main/src/osd/ECBackend.h#L211-L217

If we are going to change these around I think we should perhaps do it at the same time for both functions in another PR.  I did add documentation for this function in the header though.

File: src/erasure-code/ErasureCode.cc [hunk 6]
```diff
@@ -332,17 +332,40 @@ int ErasureCode::decode_concat(const map<int, bufferlist> &chunks,
 			       bufferlist *decoded)
 {
   set<int> want_to_read;
+  set<int> decode_chunks;
+  bool need_decode = false;
 
   for (unsigned int i = 0; i < get_data_chunk_count(); i++) {
     want_to_read.insert(chunk_index(i));
   }
+  if (chunks.size() < get_data_chunk_count()) {
```
[the remark this answers is not in the document]
Comment (markhpc) [reply src/erasure-code/ErasureCode.cc:341]: done!

File: src/erasure-code/ErasureCode.cc (hunk 3 above)
Comment (athanatos) [reply src/erasure-code/ErasureCode.cc:367]: @ronen-fr I'm not sure what you mean by a "cost function".  @markhpc I'm not sure there's actually a reason to select a non-systematic code.  Are there any non-systematic codes merged yet?  If not, I'm fine with just relying on a requirement that they be systematic and removing this method.

File: src/osd/ECBackend.cc (hunk 4 above)
Comment (athanatos) [reply src/osd/ECBackend.cc:1706]: This is among the less readable files in the OSD code base, let's not double down on my prior poor taste.  i, j, and k are fine as simple indices for short for loops, but j isn't a simple index.  It should probably be chunk_to_read.

File: src/osd/ECBackend.cc [hunk 7]
```diff
@@ -1686,6 +1686,29 @@ int ECBackend::get_min_avail_to_read_shards(
...
+  int total_chunks = (chunk_size - 1 + len) / chunk_size;
+  int first_chunk = (off / chunk_size) % data_chunk_count;
+
+  if (total_chunks > data_chunk_count) {
+    total_chunks = data_chunk_count;
+  }
+
+  for(int i = 0; i < total_chunks; i++) {
+    int j = (first_chunk + i) % data_chunk_count;
+    int chunk = (int)chunk_mapping.size()  > j ? chunk_mapping[j] : j;
```
Comment (athanatos) [inline src/osd/ECBackend.cc:1707]: Are there any plugins that actually use chunk_mapping?  The only two possibilities here based on the ErasureCodeInterface.h interface comment are (chunk_mapping.empty() || j < chunk_mapping.size()).

File: src/osd/ECBackend.cc [hunk 8]
```diff
@@ -1686,6 +1686,29 @@ int ECBackend::get_min_avail_to_read_shards(
   return 0;
 }
 
+void ECBackend::get_min_want_to_read_shards(
```
Comment (athanatos) [inline src/osd/ECBackend.cc:1689]: This duplicates some of the logic in get_want_to_read_shards (without the overload).  I'd like to see the other overload invoke this.

File: src/osd/ECBackend.cc [hunk 9]
```diff
@@ -2364,34 +2386,35 @@ struct CallClientContexts :
...
+      auto bounds = make_pair(read.get<0>(), read.get<1>());
+      auto aligned = ec->sinfo.offset_len_to_stripe_bounds(bounds);
+      if (aligned.first != read.get<0>() || aligned.second != read.get<1>()) {
+        aligned = ec->sinfo.offset_len_to_chunk_bounds(bounds);
+      }
+      ceph_assert(res.returned.front().get<0>() == aligned.first);
+      ceph_assert(res.returned.front().get<1>() == aligned.second);
       map<int, bufferlist> to_decode;
       bufferlist bl;
+
```
Comment (athanatos) [inline src/osd/ECBackend.cc:2398]: Don't introduce unrelated whitespace.

File: src/osd/ECBackend.cc [hunk 10]
```diff
@@ -2364,34 +2386,35 @@ struct CallClientContexts :
...
     for (auto &&read: to_read) {
-      pair<uint64_t, uint64_t> adjusted =
-	ec->sinfo.offset_len_to_stripe_bounds(
-	  make_pair(read.get<0>(), read.get<1>()));
-      ceph_assert(res.returned.front().get<0>() == adjusted.first);
-      ceph_assert(res.returned.front().get<1>() == adjusted.second);
+      auto bounds = make_pair(read.get<0>(), read.get<1>());
+      auto aligned = ec->sinfo.offset_len_to_stripe_bounds(bounds);
+      if (aligned.first != read.get<0>() || aligned.second != read.get<1>()) {
+        aligned = ec->sinfo.offset_len_to_chunk_bounds(bounds);
```
Comment (athanatos) [inline src/osd/ECBackend.cc:2392]: Why not align to chunk_bounds in the first place?  If offset_len_to_stripe_bounds returned the same offset and len, offset_len_to_chunk_bounds should as well as anything that is stripe aligned is also chunk aligned.

File: src/osd/ECBackend.cc [hunk 11]
```diff
@@ -2364,34 +2386,35 @@ struct CallClientContexts :
...
 	   ++j) {
 	to_decode[j->first.shard] = std::move(j->second);
       }
-      int r = ECUtil::decode(
-	ec->sinfo,
-	ec->ec_impl,
-	to_decode,
-	&bl);
+
+      int r = ECUtil::decode(ec->sinfo, ec->ec_impl, to_decode, &bl);
```
Comment (athanatos) [inline src/osd/ECBackend.cc:2406]: Why?

File: src/osd/ECBackend.cc [hunk 12]
```diff
@@ -2364,34 +2386,35 @@ struct CallClientContexts :
...
       if (r < 0) {
         res.r = r;
         goto out;
       }
+
```
Comment (athanatos) [inline src/osd/ECBackend.cc:2411]: stray whitespace

File: src/osd/ECBackend.h [hunk 13]
```diff
@@ -136,20 +136,29 @@ class ECBackend : public PGBackend {
    * ensures that we won't ever have to restart a client initiated read in
    * check_recovery_sources.
    */
+  typedef boost::tuple<uint64_t, uint64_t, uint32_t> ec_align_t;
+  typedef std::map<hobject_t,std::pair<int, extent_map>>  ec_extents_t;
```
Comment (athanatos) [inline src/osd/ECBackend.h:140]: using, not typedef

Comment (athanatos) [reply src/osd/ECBackend.h:140]: Swapping these types for their typedefs makes this PR harder to read.  If we're going to bother with this, at least ec_align_t should be a struct with useful member names.  Those changes should also be their own commit.

File: src/osd/ECBackend.h [hunk 14]
```diff
@@ -351,13 +360,15 @@ class ECBackend : public PGBackend {
...
     bool want_attrs;
     GenContext<std::pair<RecoveryMessages *, read_result_t& > &> *cb;
+    bool partial_read;
     read_request_t(
       const std::list<boost::tuple<uint64_t, uint64_t, uint32_t> > &to_read,
       const std::map<pg_shard_t, std::vector<std::pair<int, int>>> &need,
       bool want_attrs,
-      GenContext<std::pair<RecoveryMessages *, read_result_t& > &> *cb)
+      GenContext<std::pair<RecoveryMessages *, read_result_t& > &> *cb,
+      bool partial_read=false)
```
Comment (athanatos) [inline src/osd/ECBackend.h:369]: Why is this a parameter?  Wouldn't we always want to do a partial read if possible?  Why would we ever read more chunks than are necessary?

File: src/osd/ECBackend.cc [hunk 15]
```diff
@@ -2456,6 +2513,44 @@ void ECBackend::objects_read_and_reconstruct(
...
+    }
+    // Don't partial read if the EC isn't systematic
+    if (!ec_impl->is_systematic()) {
+      return false;
+    }
+    // Don't partial read if we have multiple stripes
+    if (to_read.size() != 1) {
+      return false;
+    }
+    // Only partial read if the length is inside the stripe boundary
```
Comment (athanatos) [inline src/osd/ECBackend.cc:2535]: These restrictions seem kind of brittle.  I don't see an obvious reason why the head and tail of a multi-stripe read couldn't be sub-stripe reads.  Shouldn't it simply be a function of the offsets?

**Review by athanatos — CHANGES_REQUESTED** [review]
Minor nits first, there's some stylistic stuff to clean up and some type renames that should be in a separate commit.

More substantially, there's a lot of inline offset twiddling I'd like to see tucked into well named stripe_info_t utility methods.

A major concern I have for features like this is how hard it is to actually trigger the behavior, or to inadvertently fail to trigger the behavior in testing.  This implementation has a lot of special casing for partial_read and (from what I can tell) an unnecessarily restrictive set of conditions on it.  Can we generalize it so that it basically always applies for non-stripe aligned reads?

Also, (just an fyi) I think this is going to conflict with @rzarzynski's work in https://github.com/ceph/ceph/pull/52264.

File: src/osd/ECBackend.cc (hunk 10 above)
Comment (athanatos) [reply src/osd/ECBackend.cc:2392]: Hmm, I'm actually not sure what's going on here.  to_read should be the original user offset/length, right?  res is the chunks read from the shards.  Seems like with this implementation we should be either aligning to stripe or to chunk depending on whether the read was submitted as a partial_read.  I'm not sure I understand how this conditional addresses that.

**Comment by athanatos** [comment]
The perf results are impressive, this is worth getting fixed up and merged.

File: src/erasure-code/ErasureCode.cc (hunk 3 above)
Comment (ronen-fr) [reply src/erasure-code/ErasureCode.cc:367]: @athanatos - sorry for the confusion. 'const'

**Comment by NUABO** [comment]
hi @markhpc , may I ask, will this optimization cause problems when the data is silently damaged or io error occurs? How to deal with abnormal situations, thanks

**Comment by markhpc** [comment]
@NUABO That's a good question.  AFAIK we don't do any kind of read path validation from the secondaries for primaries in replicated scenarios either.  I suppose a side benefit of the current EC scheme is that theoretically errors can be detected on read, though I confess I don't know how that plays out in practice.

**Comment by athanatos** [comment]
Bluestore stores checksums, so we're protected there at least.  The existing implementation doesn't use extra chunks to validate the data -- this isn't any less safe than what we already do.  Deep scrub will still validate against stored full object checksums.

**Comment by dvanders** [comment]
pls don't close this

File: src/osd/ECBackend.h (hunk 14 above)
Comment (rzarzynski) [reply src/osd/ECBackend.h:369]: Oops, isn't this a code code? I don't see any caller of `read_request_t()` setting it.

**Comment by xenago** [comment]
Well that's disappointing

**Comment by baergj** [comment]
@xenago This is being continued in https://github.com/ceph/ceph/pull/55196

**Comment by xenago** [comment]
> @xenago This is being continued in https://github.com/ceph/ceph/pull/55196

Thank you for linking that!

**Comment by NUABO** [comment]
> @xenago This is being continued in #55196

good news

**Comment by markhpc** [comment]
Yep!  Radek took it over and we're just waiting for QA to do some updated performance tests!
